compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -348,16 +348,30 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                             CheckInAllocMsg::OffsetFromTest,
                         )?;
 
+                        if intrinsic_name == sym::ptr_offset_from_unsigned && a_offset < b_offset {
+                            throw_ub_format!(
+                                "{} cannot compute a negative offset, but {} < {}",
+                                intrinsic_name,
+                                a_offset.bytes(),
+                                b_offset.bytes(),
+                            );
+                        }
+
                         // Compute offset.
                         let usize_layout = self.layout_of(self.tcx.types.usize)?;
                         let isize_layout = self.layout_of(self.tcx.types.isize)?;
+                        let ret_layout = if intrinsic_name == sym::ptr_offset_from {
+                            isize_layout
+                        } else {
+                            usize_layout
+                        };
                         let a_offset = ImmTy::from_uint(a_offset.bytes(), usize_layout);
                         let b_offset = ImmTy::from_uint(b_offset.bytes(), usize_layout);
                         let (val, _overflowed, _ty) =
```

Comment:
Hm, I feel like the existing code here made little sense? We want the subtraction to happen in `isize` with the signed version, right?

So this should probably all use the same layout, the one you now call `ret_layout`. (Not that it makes any difference for the result, thanks to two's complement.)

Reply:
In fact, looking at the safety docs, we should probably use `isize` for `a_offset` and `b_offset`, and then actively check for overflow here?

Reply:
That's true.  Probably both versions should always do the subtraction in `isize`, so the overflow flag will check the `isize::MAX` limit.

I can make that change, but is there a way I can write a test for it?  I would guess that CTFE won't actually let me make an allocation big enough to overflow the limit without failing the "different allocations" check first.